draft-yasskin-dispatch-web-packaging.md

```diff
@@ -0,0 +1,650 @@
+[^ednote-critical-sections]{:source="jyasskin"}
+
+[^ednote-critical-sections]: Do we need to mark critical section names?
+
+Let *index* be the result of parsing the bytes starting at offset
+*sections-start* + `section-offsets`\["indexed-content"] using the instructions in
+{{index}}.
+
+If `section-offsets`\["manifest"] is present, let *manifest* be the
+result of parsing the bytes starting at offset *sections-start* +
+`section-offsets`\["manifest"] using the instructions in {{manifest}}.
+
+The parser MAY return a semantic package consisting of *index*, and,
+if initialized, *manifest*.
+
+To parse each resource described within *index*, the parser MUST follow the
+instructions in {{resource}}.
+
+## Parsing the index {#index}
+
+The main content of a package is an index of HTTP requests pointing to HTTP
+responses. These request/response pairs hold the manifests of sub-packages and
+the resources in the package and all of its sub-packages. Both the requests and
+responses can appear in any order, usually chosen to optimize loading while the
+package is streamed.
+
+To parse the index, starting at offset *index-start*, the parser MUST do the
+following:
+
+If the byte at *index-start* is not 0x82 (the {{CBOR}} header for a 2-element
+array), the parser MUST fail.
+
+Load a CBOR item starting at *index-start* + 1 as the `index` array in the following CDDL:
+
+~~~~~ cddl
+index = [* [resource-key: http-headers,
+            offset: uint,
+            ? length: uint] ]
+
+; http-headers is a byte string in HPACK format (RFC7541).
+; The dynamic table begins empty for each instance of
+; http-headers.
+http-headers = bstr
+~~~~~
+
+If the item doesn't match this CDDL, or it is not a Canonical CBOR item (Section
+3.9 of {{CBOR}}), the parser MUST fail.
+
+Let *resources-start* be the offset immediately after the `index` item. For
+example, if *index-start* were 75 and the `index` item were 105 bytes long,
+*resources-start* would be 75+1+105=181. (1 for the 0x82 array header.)
+
+Decode all of the `resource-key`s using {{!HPACK=RFC7541}}, with an
+initially-empty dynamic table for each one. [^ednote-compression]{:
+source="jyasskin"} The decoded `resource-key`s are header lists
+({{HPACK}}, Section 1.3), ordered lists of name-value pairs.
+
+The parser MUST fail if any of the following is true:
+
+1. HPACK decoding encountered an error.
+1. Any `resource-key`'s first three headers are not named ":scheme",
+   ":authority", and ":path", in that order. Note that ":method" is
+   intentionally omitted because only the GET method is meaningful.
+1. Any of the pseudo-headers' values violates a requirement in Section
+   8.1.2.3 of {{!HTTP2=RFC7540}}.
+1. Any `resource-key` has a non-pseudo-header name that includes the
+   ":" character or is not lower-case ascii ({{HTTP2}}, Section
+   8.1.2).
+1. Any two decoded `resource-key`s are the same. Note that header
+   lists with the same header fields in a different order are not the
```

Comment:
Yes! That'll appear in the W3C spec about browsers loading packages, but basically, we're not going to (can't) locate resources by looking for an exact match between the headers the browser would generate and the key in the index. Instead, the browser will have to use its request information to pick the best of the options provided by the package.